pymeasure/instruments/instrument.py

```diff
@@ -244,36 +246,94 @@ def id(self):
 
-    def read(self):
-        """ Reads from the instrument through the adapter and returns the
-        response.
+        :param int count: Number of bytes to read. A value of -1 indicates to
+            read the whole read buffer.
+        :param kwargs: Keyword arguments for the adapter.
+        :returns bytes: Bytes response of the instrument (including termination).
         """
-        return self.adapter.read()
+        return self.adapter.read_bytes(count, **kwargs)
 
-    def values(self, command, **kwargs):
-        """ Reads a set of values from the instrument through the adapter,
-        passing on any key-word arguments.
+    # Communication functions
+    def wait_until_read(self, query_delay=0):
+        """Wait some time until reading.
+
+        :param query_delay: Delay between writing and reading in seconds.
         """
-        return self.adapter.values(command, **kwargs)
+        if query_delay:
+            time.sleep(query_delay)
 
-    def binary_values(self, command, header_bytes=0, dtype=np.float32):
-        return self.adapter.binary_values(command, header_bytes, dtype)
+    def ask(self, command, query_delay=0):
+        """ Writes the command to the instrument through the adapter
+        and returns the read response.
+
+        :param command: Command string to be sent to the instrument.
+        :param query_delay: Delay between writing and reading in seconds.
+        :returns: String returned by the device without read_termination.
+        """
+        self.write(command)
+        self.wait_until_read(query_delay)
+        return self.read()
+
+    def values(self, command, separator=',', cast=float, preprocess_reply=None):
+        """ Writes a command to the instrument and returns a list of formatted
+        values from the result.
+
+        :param command: SCPI command to be sent to the instrument
+        :param separator: A separator character to split the string into a list
+        :param cast: A type to cast the result
+        :param preprocess_reply: optional callable used to preprocess values
+            received from the instrument. The callable returns the processed
+            string.
+        :returns: A list of the desired type, or strings where the casting fails
+        """
+        results = str(self.ask(command)).strip()
+        if callable(preprocess_reply):
+            results = preprocess_reply(results)
+        results = results.split(separator)
+        for i, result in enumerate(results):
+            try:
+                if cast == bool:
+                    # Need to cast to float first since results are usually
+                    # strings and bool of a non-empty string is always True
+                    results[i] = bool(float(result))
+                else:
+                    results[i] = cast(result)
+            except Exception:
+                pass  # Keep as string
+        return results
+
+    def binary_values(self, command, query_delay=0, **kwargs):
```

Comment:
I am sorry, if this was already discussed, but why the `write_binary_values` is not provided here but instead in the `Adapter` class.

A second option, could be to define an `SCPIInstrument` class derived from `Instrument` which provides:
- `binary_values`
- `write_binary_values`
- all the basic SCPI commands implemented in `Instrument`
In class `Instrument`, remove the SCPI parameter and all implemented instrument with `SCPI` set to `True` should be modified to inherit from `SCPIInstrument` and all instrument with `SCPI` set to `False` should have the parameter removed.
Possibly this second option is a bit too much for this already huge PR.

Reply:
Dividing the instrument in two classes is a point of discussion in other issues / discussion. I think it would be good, but, as you said, not in this PR.

The write binary values is in the Adapter, because some adapters (Prologix) need to do some adjustments to the binary data (escaping some bytes) before sending it.

Reply:
Indeed, the `write_binary_values` should call the corresponding adapter method, so you don't have to worry about whether it's VISAAdapter or PrologixAdapter. Make sense?

Reply:
Ah ok, it's fine with me to add a method which calls the adapter.
First I thought you wanted me to move the whole logic to the instrument, which is not possible due to the aforementioned difficulty.

I'll make a write binariy values method.